// postcss/lib/input.js
'use strict';

let { isClean, my } = require('./symbols');
let { default: maybeParse } = require('./parse');

let fromOffsetCache = Symbol('fromOffsetCache');

class Input {
  constructor(css, opts = {}) {
    if (
      css === null ||
      typeof css === 'undefined' ||
      (typeof css === 'object' && css.raw && css.input)
    ) {
      css = css ? css.raw : css;
    }

    this.css = css.toString();

    if (this.css[0] === '\uFEFF' || this.css[0] === '\u{feff}') {
      this.css = this.css.slice(1);
    }

    this.hasBOM = this.css.length !== css.length;
    this.file = opts.from ? opts.from : undefined;
    this.id = this.file || '<input css ' + this.css.length + ' bytes>';
    this.map = opts.map ? opts.map : undefined;
    if (this.map && !this.map.file) {
      if (this.map.annotation) this.map.file = this.from;
      else this.map.file = this.id;
    }
    this.from = this.file || opts.from;
    this.source = {
      input: this,
      start: { column: 1, line: 1, offset: 0 },
    };

    if (this.map && (this.map.inline || this.map.annotation)) {
      let uri = this.map.annotation.trim();
      if (uri.slice(0, 3) === 'url') uri = uri.slice(4, -1).replace(/["']/g, '');
      if (uri !== this.map.file) {
        let root = this.map.root || process.cwd();
        let dirname = require('path').dirname;
        let from = this.map.file || opts.from;
        let to = require('path').resolve(root, uri);
        let map = require('fs').readFileSync(to, 'utf8');
        map = JSON.parse(map);
        map.file = from;
        if (root !== process.cwd()) map.root = root;
        this.map = map;
      }
    }
  }

  error(message, line, column, opts = {}) {
    let endLine, endColumn, startLine, startColumn, endOffset, startOffset;

    if (line && typeof line === 'object') {
      opts = line;
      line = undefined;
    }

    if (opts.plugin) message = '[' + opts.plugin + '] ' + message;

    if (typeof line === 'undefined' || typeof column === 'undefined') {
      let pos = this.fromOffset(0);
      line = pos.line;
      column = pos.column;
    }

    let origin = this.origin(line, column);
    let css = this.css;

    if (origin) {
      if (origin.file) {
        css = this.relative(origin.file);
      }
      line = origin.line;
      column = origin.column;
    }

    let lines = css.split(/\r?\n/);
    let start = Math.max(line - 3, 0);
    let end = Math.min(line + 2, lines.length);
    let index = Array.from({ length: end - start }, (_, i) => start + i + 1)
      .map((i) => i + ' | ' + lines[i - 1])
      .join('\n');

    let plugin = opts.plugin ? ' from ' + opts.plugin : '';

    let err = new CssSyntaxError(
      message,
      lines,
      line,
      column,
      this.source,
      this.file,
      index,
      plugin
    );

    err.input = { line, column, endLine, endColumn, endOffset, startOffset };

    return err;
  }

  get fromOffset() {
    return this[fromOffsetCache] || (this[fromOffsetCache] = this.css.split('\n'));
  }

  origin(line, column) {
    if (!this.map) return false;
    let consumer = this.map.consumer();
    let position = consumer.originalPositionFor({ line, column });
    if (!position.source) return false;

    let source = position.source;
    if (source === '<no source>') return false;

    let from = this.map.file || this.from;
    let file = require('path').resolve(from, source);
    return {
      file,
      line: position.line,
      column: position.column,
    };
  }

  toString() {
    return this.css;
  }

  [isClean]() {
    return false;
  }

  [my]() {
    return this;
  }
}

module.exports = Input;
module.exports.default = Input;
module.exports.Input = Input;
